cpp/src/arrow/compute/kernels/vector_replace.cc

```diff
@@ -442,23 +442,460 @@ struct ReplaceWithMaskFunctor {
+      case Datum::ARRAY: {
+        auto array_input = *batch[0].array();
+        bool has_current_value = false;
+        int64_t has_current_offset = -1;
+        return FillForwardArray(ctx, array_input, out, array_input, &has_current_value,
+                                &has_current_offset);
+      }
+      case Datum::CHUNKED_ARRAY: {
+        return FillForwardChunkedArray(ctx, batch[0].chunked_array(), out);
+      }
+      default:
+        break;
+    }
+    return Status::NotImplemented(
+        "Unsupported types for drop_null operation: "
+        "values=",
+        batch[0].ToString());
+  }
+
+  static Status FillForwardArray(KernelContext* ctx, ArrayData& array, Datum* out,
+                                 ArrayData current_value_array, bool* has_current_value,
+                                 int64_t* current_value_offset) {
+    ArrayData* output = out->array().get();
+    if (!output->buffers[0]) {
+      ARROW_ASSIGN_OR_RAISE(output->buffers[0], ctx->AllocateBitmap(array.length));
+    }
+    output->length = array.length;
+    int8_t direction = 1;
+
+    if (array.MayHaveNulls()) {
+      return FillNullExecutor<Type>::ExecFillNull(
+          ctx, array, array.buffers[0]->data(), output, direction, current_value_array,
+          has_current_value, current_value_offset);
+    } else {
+      *current_value_offset = FillNullExecutor<Type>::LastElementOffset(array, direction);
+      if (array.length > 0) {
+        *has_current_value = true;
+      }
+      *output = array;
+    }
+    return Status::OK();
+  }
+
+  static Status FillForwardChunkedArray(KernelContext* ctx,
+                                        const std::shared_ptr<ChunkedArray>& values,
+                                        Datum* out) {
+    if (values->null_count() == 0) {
+      *out = Datum(*ToResult(values));
+      return Status::OK();
+    }
+    if (values->null_count() == values->length()) {
+      *out = Datum(*ToResult(values));
+      return Status::OK();
+    }
+    std::vector<std::shared_ptr<Array>> new_chunks;
+
+    if (values->length() > 0) {
+      ArrayData array_with_current = *values->chunk(/*first_chunk=*/0)->data();
+      bool has_current_value = false;
+      int64_t has_current_offset = -1;
+      for (const auto& chunk : values->chunks()) {
+        auto buffer_size = chunk->length() * bit_width(values->type()->id());
+        ARROW_ASSIGN_OR_RAISE(auto data, AllocateBuffer(buffer_size, ctx->memory_pool()));
+
+        std::unique_ptr<ArrayBuilder> builder;
+        RETURN_NOT_OK(MakeBuilder(ctx->memory_pool(), values->type(), &builder));
+        RETURN_NOT_OK(builder->Reserve(chunk->length()));
+        ARROW_ASSIGN_OR_RAISE(auto array_output, builder->Finish());
+
+        ARROW_ASSIGN_OR_RAISE(array_output->data()->buffers[1],
+                              ctx->Allocate(buffer_size / 8));
+
+        array_output->data()->length = values->length();
+        auto array_out = Datum(*array_output);
```

Comment:
Instead of this, we can adjust `FillNullInDirectionImpl` to do the allocation (since it knows exactly how big the buffers need to be) and then we can specify that these kernels do not preallocate any data. (Again, we will have to separate the registration for these kernels from the registration for ReplaceWithMask.)